### PR TITLE
Cache based on timestamp and implied ordering.

### PR DESCRIPTION
**Pull request by scottsdevelopment**
I've tried my best to outline the production scenario using test cases for the proposed change. There might be a better way to architect this that you would prefer, and I am open to feedback and suggestions.

**Problem:** 

When running multiple Pipelines with the same context and reintroducing the same candle from a source, the cache will append duplicates causing problems for the rest of the calculations.

Additionally, a live source may emit the same candle multiple times until it is completed, however we can handle these scenarios in additional code changes or after we work out how we want things to behave.

**Consider the solution below:**

- Candles are loaded from the database and cached, enough so that new indicators may be calculated. This is Pipeline 1.
- `LastSymbolTimestamp` is the earliest timestamp for the series of symbols provided. Since currently Sources share the same `time_from` and `time_to` across symbols, this seemed appropriate for logic.
- Using `LastSymbolTimestamp` we calculate a new `from_time` for our second `Source`, perhaps slightly adjusted for any recent corrections, we can't guarantee there won't be overlapping candles. 
- Pipeline 2 is demonstrated to run technical processors from a ListSource on just enough data to get our technicals running on the next candle.
- Pipeline 3 loads live data using the delta.
- In summary, old data is loaded first, context is built, new data is then loaded, processed, and then synced. Performance is greatly improved.

I've also included my example MongoDB aggregator to normalize the existing data prior to the changes and squash any duplicated data that may have been introduced, this was super useful to preserve minute candles as these are often limited history windows.

**Summary of changes:**

- Update `CandleCache` to use an explicit ordered dictionary to hold candle values by timestamp keys. 
- Move `FakeSource` to `ListSource` as a main source component.
- Create `ContextInjectionFactorySource` as a new source type to act as an intermediate between the construction of a source and the modification of the context.
- Add `LastSymbolTimestamp` terminator, which holds the lowest last timestamp from all symbols provided by reading cache at termination and storing a cache value.
- Add script example for MongoDB to normalize and merge duplicate candles from previous change.
- Update and provide unit tests where acceptable.

**Once we discuss this proposed PR, there may be additional work items required before submission:**
- Consider removing cast to list on the `get_symbol_candles`.
- Consider using key checks for `reprocess`
- Any dependent changes that cascade from these changes.

~~Code snippet updated below~~

### CODE REVIEW

**Comment by scottsdevelopment** [comment]
Note for serialization, I was exploring how this could be done with lambdas and I found this library which extends the pickle library, https://pypi.org/project/cloudpickle/ , but random code execution in json is also scary, so I'm not sure how this should work.

**Comment by idanya** [comment]
> **Problem:**
> 
> When running multiple Pipelines with the same context and reintroducing the same candle from a source, the cache will append duplicates causing problems for the rest of the calculations.

When running multiple pipelines, do you really need a source on the 2nd pipeline? If the 1st and 2nd pipelines are running different processors on the same input data, can't the processors run one after the other on the same pipeline, saving the complete results to DB? 

> 
> Additionally, a live source may emit the same candle multiple times until it is completed, however we can handle these scenarios in additional code changes or after we work out how we want things to behave.
> 

Sounds like a specific data provider issue that may be handled on the Source/Provider side. There is also the `Events` mechanism that might mitigate these kinds of sync issues. For example, firing an even when a `TimeSpan` interval has passed.

> **Consider the solution below:**
> 
> * Candles are loaded from the database and cached, enough so that new indicators may be calculated. This is Pipeline 1.

as said above, what is the motivation to end the pipeline and not continue this one with the 2nd pipeline processors?

That said, I think `ContextInjectionFactorySource`  can be a helpful component to base a `Source` of one pipeline on the results of an earlier one. Trying to think if we can have a simpler way to achieve the same result while keeping it dynamic, as creating a factory with a custom delegate can be harder to understand and implement.

**Comment by idanya** [comment]
> Note for serialization, I was exploring how this could be done with lambdas and I found this library which extends the pickle library, https://pypi.org/project/cloudpickle/ , but arbitrary code execution from json is also scary, so I'm not sure how this should work.

One of the ideas behind serialization is to enable the ability to offload the creation of components to 3rd party. This is planned to be the facade for a GUI/CLI and this is why serialization was implemented in a known, human-readable, and writeable format.

**Comment by scottsdevelopment** [comment]
> When running multiple pipelines, do you really need a source on the 2nd pipeline? If the 1st and 2nd pipelines are running different processors on the same input data, can't the processors run one after the other on the same pipeline, saving the complete results to DB?
> as said above, what is the motivation to end the pipeline and not continue this one with the 2nd pipeline processors?

For backstory, 

[Initially the `PipelineRunner` was called twice](https://github.com/idanya/algo-trader/pull/18/commits/2cc382903f4f24623d3198b40eb08b6ed7e959b4#diff-f0410419e38532d5fe1a3f945cd75ac927206f3f593e4ca13177ead573cf8eb3) to pass along the shared context before realizing a factory pattern would allow me to maintain a singular runner and use a pipeline list.

I ran into speed issues where I didn't want to process the 14,000 records from my MongoDB through the TA system and resink, but I also wanted them loaded in cache with enough TA to generate the next incoming candle and save those, I don't actually have a valid use-case for the other 99% of the candles, so I believe I did over engineer the `pipeline_two` (hence known as the `ListSource` pipeline) pipeline. 

Rethinking this real-world example, I should be thinking about using a sane initial `from_time` for `pipeline_one` (the MongoDB source). Say for only the last 3 hours' worth of minute candles, or however much cache and technical would actually be needed in the later pipeline/processors. This would eliminate the ListSource pipeline from the example. In `pipeline_three` (the Live Source) we still get overlap on the candle cache, the source doesn't know what data is cached, so these changes seemed required. 

The main point of this PR is to validate that in the current landscape, caching a candle by timestamp, doesn't result in duplicate entries. There was some consideration I made about [symbol][timespan][timestamp] cache lookup path, however it became use-case scope creep and I decided again it's something that if I had need for could be adjusted in another PR. When gathering new data after loading previous data, I am not easily able to guarantee there won't be old candles even when our source knows the ending time of our last candle. Is there a better way to achieve this all together?


Thanks for the feedback, I will have some of the examples and code adjusted when I get a chance. 

> Sounds like a specific data provider issue that may be handled on the Source/Provider side. There is also the `Events` mechanism that might mitigate these kinds of sync issues. For example, firing an even when a `TimeSpan` interval has passed.

I'm not worried about this specific case quite yet, it will be something for me to think about, and these are good ideas.

File: src/pipeline/terminators/last_candle_timestamp.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+class LastSymbolTimestamp(Terminator):
+    def __init__(self, symbols: List[str]) -> None:
+        super().__init__()
+        self.symbols = symbols
+
+    def terminate(self, context: SharedContext):
+        cache_reader = CandleCache.context_reader(context)
+        last_symbol_timestamp = datetime.now()
+        for symbol in self.symbols:
+            last_candle_timestamp = cache_reader.get_symbol_candles(symbol)[-1].timestamp
```
Comment (scottsdevelopment) [inline src/pipeline/terminators/last_candle_timestamp.py:20]: If there are no candles, this would cause an error.

**Comment by scottsdevelopment** [comment]
Thinking about my question out loud, we could only emit candles that satisfy the condition of being past the initial `from_time`. That might eliminate the changes to the cache all together, leaving the context factory injection source, I would have to test a bit more tomorrow. 

Thanks for the discussion.

**Comment by scottsdevelopment** [comment]
After looking at real world use case again, the optimization ended up being something like this:

```python
    mongodb_storage = MongoDBStorage()
    candles = []
    
    for symbol in symbols:
        candles += [mongodb_storage._deserialize_candle(candle) for candle in mongodb_storage.candles_collection.find({
            'symbol': symbol,
            'timespan': TimeSpan.Minute.value,
        }).limit(TimeSpan.Minute.value * 5).sort("timestamp", -1)]
    source = ListSource(candles)
    processors = TechnicalsProcessor(TechnicalsNormalizerProcessor(next_processor = CandleCache()))
    terminator = LastSymbolTimestamp(symbols)
    pipeline_one = Pipeline(source, processors, terminator)
    
    # Switched to IB for example
    live_source = ContextInjectionFactorySource(context, lambda context: IBHistorySource(InteractiveBrokersConnector(), symbols, TimeSpan.Minute, LastSymbolTimestamp.get(context)))
    sink = MongoDBSinkProcessor(mongodb_storage)
    cache_processor = CandleCache(sink)
    technical_normalizer = TechnicalsNormalizerProcessor(next_processor = cache_processor)
    technicals = TechnicalsProcessor(technical_normalizer)
    
    pipeline_two = Pipeline(live_source, technicals)
    
    PipelineRunner([pipeline_one, pipeline_two], context).run()  
  ```
  
I still run into an issue where the last candles from the DB may be cached, however we want to recache them with updated data, which would put the candles emitted after an adjusted `time_from`, but still duplicating in the cache. This usually occurs on the last candle that was being reported, or if there was an adjustment from the provider when recollecting the data.

Any thoughts on this?

**Comment by idanya** [comment]
I read your use case and understand the issues. I'll give you a few pointers that I think might help clear things out.

1. Instead of having the whole `ContextInjectionFactorySource` mechanism, we could think about just passing the context to the source run() function as the pipeline does with processors. This will let the Source the option to decide where it should start (in your case) or what other configurations it should you. 
2. I see you reading from MongoDB and appending candles to `ListSource` - You have `MongoDBSource` for that. Maybe you should consider just writing a new Mongo Source that fits your needs. 
3. Let's remember that `CandleCache` at the end of the day is a utility for accessing the context. You don't have to change it for this particular use case, you could just create another processor that fits your needs (`UniqueCandleCache` (?) ). In that specific implementation you could decide that, for example, if you get a candle you already have, you overwrite the old one and just not call the next processor, or add an attachment to the candle that says "do everything TA, but don't trade upon it".

**Comment by scottsdevelopment** [comment]
> I read your use case and understand the issues. I'll give you a few pointers that I think might help clear things out.
> 
> 1. Instead of having the whole `ContextInjectionFactorySource` mechanism, we could think about just passing the context to the source run() function as the pipeline does with processors. This will let the Source the option to decide where it should start (in your case) or what other configurations it should you.

Can you elaborate on `Source.run(context)` method that should be called when the pipeline starts, would it also make sense for `Processor` to also have a `Processor.run(context)`?  Could you read over issue #12 while we are on this subject? This kind of re-architecture could be phased, but I think it should be considered as part of the bigger plan.

> 2. I see you reading from MongoDB and appending candles to `ListSource` - You have `MongoDBSource` for that. Maybe you should consider just writing a new Mongo Source that fits your needs.

I understood that, but would it make sense to modify `MongoDBSource` to support modes and argument options, or just keep adding different MongoDBSource? I feel like there are pros and cons to both. 

> 3. Let's remember that `CandleCache` at the end of the day is a utility for accessing the context. You don't have to change it for this particular use case, you could just create another processor that fits your needs (`UniqueCandleCache` (?) ). In that specific implementation you could decide that, for example, if you get a candle you already have, you overwrite the old one and just not call the next processor, or add an attachment to the candle that says "do everything TA, but don't trade upon it".

The `CandleCacheContextReader` is highly coupled with most of the provided processors, so I think creating another cache processor would work well, if we used the same identity key, or we would have to adjust processors to determine which cache to read. Another option is to again, extend the current `CandleCache` to support caching modes and options. I also think that when seeing a candle being added to the cache again, it would be a valid candidate to use `reprocess`.

I would like to hold off on the PR until we determine again, what the total scope of the changes are, but I did need to have the issues present so we could have this discussion.

**Comment by idanya** [comment]
Sorry, I meant Source.read(). So that when the second pipeline will start reading its source, it will get the context from the former pipeline. This answers the `MongoDBSource` question as well, so that as a Source, he can read parameters that will guide it on how to use the db.

The `CandleCacheContextReader` is basically an adapter to get a typed struct from the cache.  An implementation of 
 a different `CandleCache` can use whatever key it wants (see `CONTEXT_IDENT`) and would not have to overwrite other processors.

**Comment by scottsdevelopment** [comment]
> Sorry, I meant Source.read(). So that when the second pipeline will start reading its source, it will get the context from the former pipeline. 

Ok. I will look into this when I get a chance and see what comes of it.

> This answers the `MongoDBSource` question as well, so that as a Source, he can read parameters that will guide it on how to use the db.

So just to confirm, we are suggesting extending the current `MongoDBSource` to support read mode for "last x candles by timestamp over symbols", would we also be extending the `StorageProvider`?

> The `CandleCacheContextReader` is basically an adapter to get a typed struct from the cache. An implementation of a different `CandleCache` can use whatever key it wants (see `CONTEXT_IDENT`) and would not have to overwrite other processors.

Your suggestion implies that anything that calls `CandleCache.context_reader` would need to be refactored to supply a cache key type, and the logic around the `symbol_candles` not always needing to append the last candle would need to change. So, either we abstract parts of `TechnicalProcessor` (and all the other ones too, since most processors use this code pattern), decouple the dependency on `CandleCache` everywhere, or support logic for cache modes (Append, Unique) as a requirement. 

I'm in favor of supporting logic for multiple MongoDB read modes, and multiple cache modes as the PR already achieves this to some extent with a few more parts required. 

Maybe it makes sense to use method abstraction and class inheritance and provide multiple processors, but I think it would make the API clunky, I would prefer to not introduce multiple classes that do the same things albeit slightly different and instead provide modes or logic to support different setups and decide which setups are worth supporting as a requirement.

Can you please address the concerns I have raised about the processors and the cache method? The current PR starts to address them, and my next commit has more fixes and corrections.

Let me continue with some additional commits and we can see how things go?

File: src/pipeline/processors/technicals.py [hunk 2]
```diff
@@ -32,8 +32,9 @@ def process(self, context: SharedContext, candle: Candle):
         cache_reader = CandleCache.context_reader(context)
         symbol_candles = cache_reader.get_symbol_candles(candle.symbol) or []
 
-        if candle.timestamp in list(map(lambda candle: candle.timestamp, symbol_candles)):
-            candles = symbol_candles
+        candle_index = [x for x in range(len(symbol_candles)) if symbol_candles[x].timestamp == candle.timestamp] or None
+        if candle_index and candle_index[0] > 0:
```
Comment (scottsdevelopment) [inline src/pipeline/processors/technicals.py:36]: There might be an issue when candle_index[0] is 0, review the logic, perhaps add supporting tests if this is the way we go.

**Comment by idanya** [comment]
1. We do not necessarily need to change `MongoDBSource`. It utilizes the `MongoDBStorage`, so a different (new)`Source` that fits your needs can be implemented. 
2. Same goes for `CandleCache`. It's a processor with its preferred way and structure to save data in to cache. Any other processor can decide how and when to save data to the context.

**Comment by scottsdevelopment** [comment]
Ok steps for the PR.

1. Inject context into Source.read.

2. Provide LastCandleTimestampSource which will modify the configurable
from_time on the inner source before read.

3. Add additional Mongo Source for new read method.

4. Add additional Cache Processor for different cache method and key.

> Any other processor can decide how and when to save data to the context.

5. Update other provided processors to support multiple cache processor.


@idanya,

If you would provide some specific feedback on refactoring of the Technical
processor to support changes 5 based off the current PR, it would be
helpful.

Specifically,

1. I'm trying to determine the best way to handle the dependant static
method CandleCache.get_reader to determine which cache to read during
process.

2. I've also added some additional logic after the get reader but in this
PR, any feedback or concerns about this approach.

**Comment by idanya** [comment]
1. yes
2. this is for your specific implementation
3. also for your specific implementation
4. yes
5. Processors don't accept CandleCache. it's invisible to them. They accept the context which is dynamic k/v store.  

Currently, there is no space/directory to place your own processors/sources/etc... implementations. 
I guess it would be good to put an empty directory as a placeholder for private specific user implementations and put it on .gitignore

**Comment by scottsdevelopment** [comment]
I think what you are suggesting is to just create a space for independent
code and not worry about this issue, except to change some architecture or
introduce generic code to support the independent code better. I am okay
with that.

I have some ideas on how to proceed now.

**Comment by idanya** [comment]
Basically yes. 
I do think passing the context to the source.read() is a good idea that can serv the border use-case.

**Comment by idanya** [comment]
@scottsdevelopment Should we close this PR?

**Comment by scottsdevelopment** [comment]
Let's close this, I will reopen a PR with our suggestions, just have been busy since holidays.

**Pull request closed without merging**